Cabal-syntax/src/Distribution/Types/PackageDescription/Lens.hs

```diff
@@ -46,7 +47,7 @@ licenseRaw :: Lens' PackageDescription (Either SPDX.License License)
 licenseRaw f s = fmap (\x -> s{T.licenseRaw = x}) (f (T.licenseRaw s))
 {-# INLINE licenseRaw #-}
 
-licenseFiles :: Lens' PackageDescription [SymbolicPath PackageDir LicenseFile]
+licenseFiles :: Lens' PackageDescription [RelativePath Pkg File]
```

Comment:
I'm very unhappy to see these kind of changes. You break literally all `Cabal-syntax` consumers, and you don't get any benefit. The `SymbolicPath PackageDir LicenseFile` was already *relative*, it's a (relative) path from package directory to the license file.

You also drop some type-safety, as `LicenseFile` was telling that it's not *some* file, but a *license* file (e.g. not a some file in `extraSrcFiles`.

Again, I'm very unhappy to see this. You should have asked before accepting 10k diff.

Reply:
In particular, there shouldn't be any *non-relative* files in `.cabal` files. The offenders like `extra-lib-dirs` simply don't belong into package files; those should be configured by users externally (e.g. in `cabal.project` or `cabal.config` files). That's the historical mistake which should be cleaned up, not further cemented.

Reply:
Forcing users who want to install a package to turn it into a project or otherwise hack their config to make it work on e.g. MacOS isn't really acceptable.

Reply:
I didn't mean you need to remove the support for `extra-lib-dirs` right away, but having system specific hard-coded paths is not correct either. The absolute paths do not belong into package definitions. That's why there are `build-type: Configure`, (soon `Hooks`) and `build-type: Custom`.

Reply:
> I'm very unhappy to see these kind of changes. You don't get any benefit. The `SymbolicPath PackageDir LicenseFile` was already _relative_, it's a (relative) path from package directory to the license file.
> 
> You also drop some type-safety, as `LicenseFile` was telling that it's not _some_ file, but a _license_ file (e.g. not a some file in `extraSrcFiles`.
> 
> Again, I'm very unhappy to see this. You should have asked before accepting 10k diff.

Oleg, I'm sorry but I spent a lot of time trying to come up with a design that was workable for the whole of Cabal. To do that, I did find it necessary to change the API in several ways. For instance, yes, we no longer specify the exact kind of file:

  - it lead to a proliferation of definitions for all the different kinds of files,
  - in my experience it was not useful to keep track of which specific file something was pointing to, it was extra book-keeping that didn't in practice avoid any bugs.
  
One other major point of departure came from the fact I realised we unfortunately would not be able to only use relative symbolic paths everywhere; there are a lot of parts in Cabal which use paths that are either relative or absolute, so I had to make that work with the API.

I understand if you think this "cemented" bad parts of the codebase, but I want to say that it is much easier now to change these things now because it is often a case of doing a search/replace, whereas when most of the codebase was "untyped" (= using `FilePath`) it took a lot of effort to identify the places which needed to be changed.

I know and can tell you put serious consideration into the design of the initial API. I honestly did try sticking to it as much as I could but I was also informed by looking at what was actually in use in the Cabal codebase. Maybe you think that's a mistake, and that I should have stuck to a design and tried harder to make Cabal adapt to that abstract ideal, but I don't think I would have managed to write that patch with that approach.

Reply:
>  there are a lot of parts in Cabal which use paths 

That is the main pain point.

The `SymbolicPath` was intended to be used as *symbolic path*, i.e. paths to be interpreted that appear in `.cabal` files. Not any real filepath which may appear on the filesystem. That should had different type. `FilePath` or `OsPath` or ... I.e. not as typed wrapper around `filepath` / `directory`. Maybe it's not the case, but the distinction is now very blurry. (In particular, `CWD` root doesn't make sense for `SymbolicPath`s as paths in `.cabal` file in my design).

TL;DR I hope that `SymbolicPath` is not the only abstraction over paths, as in my opinion, it simply won't work well. Too many concerns, too complicated API.